esda/shape.py

```diff
@@ -0,0 +1,332 @@
+
+    See the Notes for information on the shape of angles and indices.
+
+    Notes
+    -------
+    If a geometry has n coordinates and k parts, the array will be n - k.
+    If each geometry has n_i coordinates, then let N be a vector storing
+    those counts (computed, for example, using pygeos.get_num_coordinates(ga)).
+    Likewise, let K be a vector storing the number of parts each geometry has, k_i
+    (computed, for example, using pygeos.get_num_geometries(ga))
+
+    Then, the output is of shape (N - K).sum()
+
+    """
+    exploded = pygeos.get_parts(ga)
+    coords = pygeos.get_coordinates(exploded)
+    n_coords_per_geom = pygeos.get_num_coordinates(exploded)
+    n_parts_per_geom = pygeos.get_num_geometries(exploded)
+    angles = numpy.asarray(_get_angles(coords, n_coords_per_geom, n_parts_per_geom))
+    if return_indices:
+        return angles, numpy.repeat(
+            numpy.arange(len(ga)),
+            pygeos.get_num_coordinates(ga) - pygeos.get_num_geometries(ga),
+        )
+    else:
+        return angles
+
+
+@njit
+def _get_angles(points, n_coords_per_geom, n_parts_per_geom):
+    """
+    Iterate over points in a set of geometries.
+    This assumes that the input geometries are simple, not multi!
+    """
+    offset = int(0)
+    start = points[0]
+    on_geom = 0
+    on_coord = 0
+    result = []
+    n_points = len(points)
+    while True:
+        if on_coord == (n_coords_per_geom[on_geom] - 1):
+            offset += on_coord
+            on_geom += 1
+            on_coord = 0
+            if on_geom == len(n_coords_per_geom):
+                break
+            else:
+                continue
+        left = points[offset + on_coord % (n_coords_per_geom[on_geom] - 1)]
+        center = points[offset + (on_coord + 1) % (n_coords_per_geom[on_geom] - 1)]
+        right = points[offset + (on_coord + 2) % (n_coords_per_geom[on_geom] - 1)]
+        a = left - center
+        b = right - center
+        angle = numpy.math.atan2(a[0] * b[1] - a[1] * b[0], numpy.dot(a, b))
+        result.append(angle)
+        on_coord += 1
+    return result
+
+
+# -------------------- INTERTIAL MEASURES -------------------- #
+
+
+def isoperimetric_quotient(collection):
+    """
+    The Isoperimetric quotient, defined as the ratio of a chain's area to the
+    area of the equi-perimeter circle.
+
+    Altman's PA_1 measure
+
+    Construction:
+    --------------
+    let:
+    p_d = perimeter of district
+    a_d = area of district
+
+    a_c = area of the constructed circle
+    r = radius of constructed circle
+
+    then the relationship between the constructed radius and the district
+    perimeter is:
+    p_d = 2 \pi r
+    p_d / (2 \pi) = r
+
+    meaning the area of the circle can be expressed as:
+    a_c = \pi r^2
+    a_c = \pi (p_d / (2\pi))^2
+
+    implying finally that the IPQ is:
+
+    pp = (a_d) / (a_c) = (a_d) / ((p_d / (2*\pi))^2 * \pi) = (a_d) / (p_d**2 / (4\PI))
```

Comment:
Just a nit here, which can be ignored since it is a pain.

It would be nicer to have these formulas within a `.. math::` block.

Reply:
Yeah... I haven't yet mastered how to work with Sphinx documentation formatting yet... happy to format it however makes it look pretty online!